framework/include/executioners/PicardSolve.h

```diff
@@ -19,166 +20,76 @@ class PicardSolve;
-  Real _picard_rel_tol;
-  /// Absolute tolerance on residual norm
-  Real _picard_abs_tol;
-  /// Postprocessor value for user-defined picard convergence check
-  const PostprocessorValue * const _picard_custom_pp;
-  /// Relative tolerance on postprocessor value
-  Real _custom_rel_tol;
-  /// Absolute tolerance on postprocessor value
-  Real _custom_abs_tol;
-  /// Whether or not we force evaluation of residual norms even without multiapps
-  bool _picard_force_norms;
-  /// Relaxation factor for Picard Iteration
-  Real _relax_factor;
-  /// The transferred variables that are going to be relaxed
-  std::vector<std::string> _relaxed_vars;
-
-  /// Relaxation factor outside of Picard iteration (used as a subapp)
-  Real _picard_self_relaxation_factor;
-  /// Variables to be relaxed outside of Picard iteration (used as a subapp)
-  std::vector<std::string> _picard_self_relaxed_variables;
-
-  /// Maximum number of xfem updates per step
-  unsigned int _max_xfem_update;
-  /// Controls whether xfem should update the mesh at the beginning of the time step
-  bool _update_xfem_at_timestep_begin;
+  virtual void transformVariables(const std::set<dof_id_type> & transformed_dofs,
+                                  const bool primary) override final;
 
-private:
-  /// Timer for Picard iteration
-  const PerfID _picard_timer;
-
-  ///@{ Variables used by the Picard iteration
-  /// Picard iteration counter
-  unsigned int _picard_it;
-  /// Initial residual norm
-  Real _picard_initial_norm;
-  /// Full history of residual norm after evaluation of timestep_begin
-  std::vector<Real> _picard_timestep_begin_norm;
-  /// Full history of residual norm after evaluation of timestep_end
-  std::vector<Real> _picard_timestep_end_norm;
-  /// Status of Picard solve
-  MoosePicardConvergenceReason _picard_status;
-  ///@}
-
-  /// Counter for number of xfem updates that have been performed in the current step
-  unsigned int _xfem_update_count;
-  /// Whether step should be repeated due to xfem modifying the mesh
-  bool _xfem_repeat_step;
-
-  /// Time of previous Picard solve as a subapp
-  Real _previous_entering_time;
-
-  const std::string _solve_message;
-
-  /// Whether the user has set the auto_advance parameter for handling advancement of
-  /// sub-applications in multi-app contexts
-  const bool _auto_advance_set_by_user;
-
-  /// The value of auto_advance set by the user for handling advancement of sub-applications in
-  /// multi-app contexts
-  const bool _auto_advance_user_value;
-
-  /// force the current step to fail, triggering are repeat with a cut dt
-  bool _fail_step;
+  /// Print the convergence history of the coupling, at every coupling iteration
+  virtual void printFixedPointConvergenceHistory() override final;
+
+  /// Vector tag id for the previous solution variable, as a main app
+  TagID _old_tag_id;
+
+  /// Vector tag id for the previous solution variable, as a sub app
+  TagID _secondary_old_tag_id;
```

Comment:
Should we change "secondary" to "sub"?

Reply:
meh. should we?

Reply:
When I look at secondary/primary, I always think about it is a contact problem

But I do not feel strongly about it